rust/email/src/lib.rs

```diff
@@ -1,58 +1,142 @@
 use lettre::{
-    message::header::ContentType, transport::smtp::authentication::Credentials, Message,
-    SmtpTransport, Transport,
+	message::header::ContentType,
+	transport::smtp::authentication::Credentials, Message,
+	SmtpTransport, Transport,
 };
-use std::ffi::CStr;
-use std::os::raw::c_char;
+use std::{ffi::CStr, os::raw::c_char, slice};
 
+/// Represents the necessary information to send an email.
+/// Struct is designed to be compatible with C, allowing for easy FFI.
+#[repr(C)]
+pub struct EmailInfo {
+	from: *const c_char,
+	to: *const *const c_char,
+	to_len: usize,
+	cc: *const c_char,
+	body: *const c_char,
+
+	smtp_host: *const c_char,
+	smtp_username: *const c_char,
+	smtp_password: *const c_char,
+}
+
+/// Converts a C-style string to a Rust `String`.
+/// Returns `Ok(String)` if conversion is successful, `Err(i32)` if the input string is null.
+///
+/// # Arguments
+///
+/// * `str_` - A pointer to the null-terminated C-style string to convert.
+fn c_char_to_string(str_: *const c_char) -> Result<String, i32> {
+	unsafe {
+		if str_.is_null() {
+			return Err(-1);
+		}
+		Ok(CStr::from_ptr(str_).to_string_lossy().into_owned())
+	}
+}
+
+/// Sends an email based on the provided `EmailInfo`.
+/// Initializes logging, constructs the email message, and sends it using SMTP.
+///
+/// # Arguments
+///
+/// * `email_info` - Struct containing information about the email to be sent.
+///
+/// # Returns
+///
+/// Returns `0` on success, `-1` on failure.
+/// # Example Usage in C
+///
+/// This example demonstrates how to construct and pass an `EmailInfo` struct from C,
+/// calling the `send_email` function exposed by Rust.
+///
+/// ```c
+/// #include <stdio.h>
+/// #include "path/to/generated_header.h" // Include the generated header for Rust FFI
+///
+/// int main() {
+///     // Email information
+///     const char* from = "sender@example.com";
+///     const char* to[] = {"recipient1@example.com", "recipient2@example.com"};
+///     size_t to_len = 2; // Number of recipients
+///     const char* cc = "cc@example.com";
+///     const char* body = "Hello from C! This is the email body.";
+///	const char *smtp_host = "posteo.de";
+///	const char *smtp_username = "username";
+///	const char *smtp_password = "password";
+///
+///	// Construct the EmailInfo struct as defined in Rust, adapted for use in C
+///	EmailInfo email_info = {
+///		.from = from,
+///		.to = to,
+///		.to_len = to_len,
+///		.cc = cc,
+///		.body = body,
+///		.smtp_host = smtp_host,
+///		.smtp_username = smtp_username,
+///		.smtp_password = smtp_password,
+///	};
+///
+///     // Call the `send_email` function and capture the return value
+///     int result = send_email(email_info);
+/// ```
```

Comment:
Is it necessary to have the same example twice in our code? Maybe keep it only in email-bindings.h since thats where the `send_email` function is declared?

Reply:
It is necessary, we don't write the binding headers, `cbindgen` does it automatically.